// @nicky-lenaers/ngx-scroll-to/esm5/lib/scroll-to-animation.js
import { ReplaySubject } from 'rxjs';
import { EASING } from './scroll-to-helpers';
/** Scroll To Animation */
var ScrollToAnimation = /** @class */ (function () {
    /**
     * Class Constructor.
     *
     * @param container            The Container
     * @param listenerTarget       The Element that listens for DOM Events
     * @param isWindow             Whether or not the listener is the Window
     * @param to                   Position to scroll to
     * @param options              Additional options for scrolling
     * @param isBrowser            Whether or not execution runs in the browser
     *                              (as opposed to the server)
     */
    function ScrollToAnimation(container, listenerTarget, isWindow, to, options, isBrowser) {
        var _this = this;
        this.container = container;
        this.listenerTarget = listenerTarget;
        this.isWindow = isWindow;
        this.to = to;
        this.options = options;
        this.isBrowser = isBrowser;
        /** Recursively loop over the Scroll Animation */
        this.loop = function () {
            _this.timeLapsed += _this.tick;
            _this.percentage = (_this.timeLapsed / _this.options.duration);
            _this.percentage = (_this.percentage > 1) ? 1 : _this.percentage;
            // Position Update
            _this.position = _this.startPosition +
                ((_this.startPosition - _this.to <= 0 ? 1 : -1) *
                    _this.distance *
                    EASING[_this.options.easing](_this.percentage));
            if (_this.lastPosition !== null && _this.position === _this.lastPosition) {
                _this.stop();
            }
            else {
                _this.source$.next(_this.position);
                _this.isWindow
                    ? _this.listenerTarget.scrollTo(0, Math.floor(_this.position))
                    : _this.container.scrollTop = Math.floor(_this.position);
                _this.lastPosition = _this.position;
            }
        };
        this.tick = 16;
        this.interval = null;
        this.lastPosition = null;
        this.timeLapsed = 0;
        this.windowScrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
        if (!this.container) {
            this.startPosition = this.windowScrollTop;
        }
        else {
            this.startPosition = this.isWindow ? this.windowScrollTop : this.container.scrollTop;
        }
        // Correction for Starting Position of nested HTML Elements
        if (this.container && !this.isWindow) {
            this.to = this.to - this.container.getBoundingClientRect().top + this.startPosition;
        }
        // Set Distance
        var directionalDistance = this.startPosition - this.to;
        this.distance = this.container ? Math.abs(this.startPosition - this.to) : this.to;
        this.mappedOffset = this.options.offset;
        // Set offset from Offset Map
        if (this.isBrowser) {
            this.options
                .offsetMap
                .forEach(function (value, key) { return _this.mappedOffset = window.innerWidth > key ? value : _this.mappedOffset; });
        }
        this.distance += this.mappedOffset * (directionalDistance <= 0 ? 1 : -1);
        this.source$ = new ReplaySubject();
    }
    /**
     * Start the new Scroll Animation.
     *
     * @returns         Observable containing a number
     */
    ScrollToAnimation.prototype.start = function () {
        clearInterval(this.interval);
        this.interval = setInterval(this.loop, this.tick);
        return this.source$.asObservable();
    };
    /**
     * Stop the current Scroll Animation Loop.
     *
     * @param force          Force to stop the Animation Loop
     * @returns               Void
     */
    ScrollToAnimation.prototype.stop = function () {
        clearInterval(this.interval);
        this.interval = null;
        this.source$.complete();
    };
    return ScrollToAnimation;
}());
export { ScrollToAnimation };
